Etcd prototype files

bschatz-swift opened the pull request:
Read the README for tips

bschatz-swift committed:
First draft of etcd Vagrant files, etc

bschatz-swift committed:
Script to start etcd

bschatz-swift committed:
update token name

bschatz-swift committed:
A readme

bschatz-swift committed:
Remove old file

edmc-ss commented on etcd/Vagrantfile:
```diff
@@ -0,0 +1,24 @@
+        config.vm.define "etcd#{machine_id}" do |machine|
+            machine.vm.hostname = "etcd#{machine_id}"
+            machine.vm.network "private_network", ip: "192.168.60.#{10+machine_id}", :name => "vboxnet0", :adapter => 2
+            if machine_id == N-1
```
Not sure why this is only to be done on the "last" machine_id...

bschatz-swift replied:
I grabbed a lot of this file off the web and thought it was throw away code for now so I did not look at it in detail.

I went back and looked at it and I could not understand it completely.   I want the vagrant_bootstrap.sh script to run on all nodes to make sure that we have all updates.  I have since moved that outside the check.

We only want the ansible section to run after all nodes are configured since it verifies that etcd is up and running.   We do not want that to happen every loop

bschatz-swift committed:
Moving vagrant_bootstrap.sh outside of check

bschatz-swift committed:
Set /vagrant directory correctly

bschatz-swift committed:
Setting up etcd nodes with Go, etc

bschatz-swift committed:
Add proper path for /vagrant, install tools and unit tests

Set /vagrant to be the same as SAIO.

Install some of the same tools as SAIO (not gcc yet).

Add unit test to test basic Get()/Put() and Txn()

bschatz-swift committed:
Unit test to update multiple keys at once

bschatz-swift committed:
Small cleanup

bschatz-swift committed:
Initial prototype with node going to STARTING state

bschatz-swift committed:
Fixes for watcher and transition all nodes to ONLINE

bschatz-swift committed:
Break watcher out into multiple watch types

bschatz-swift committed:
Cleanup transaction code and change state using txn()

bschatz-swift committed:
Add unit test for HB testing

bschatz-swift committed:
Send heartbeats

bschatz-swift committed:
Add test to get multiple keys in one GET

bschatz-swift committed:
Calculate dead nodes

bschatz-swift committed:
Mark node DEAD when goes down and cleanup state when comes back

bschatz-swift committed:
Break code into multiple files

bschatz-swift committed:
Add basics of VG support

bschatz-swift committed:
Break out functions...

bschatz-swift committed:
Rearrange code and more vg work

bschatz-swift committed:
Rearrange code to make readable/modular, add VG unit tests

bschatz-swift committed:
Cleanup and document TODOs

bschatz-swift committed:
Cleanup and document TODOs

bschatz-swift committed:
Add and remove of VG with unit tests...

bschatz-swift committed:
Remove sync point which is not used

bschatz-swift committed:
Add online of VG

bschatz-swift committed:
Add infinite loop to prototype to show daemon running

Add loop to prototype to keep daemon up to simulate proxyfsd running.
Otherwise, it will come up and leave...

bschatz-swift committed:
cleanup

bschatz-swift committed:
First pass at the VG state diagram

bschatz-swift committed:
First pass at VG state diagram

bschatz-swift committed:
Move to node instead of VG

bschatz-swift committed:
Add failover of VG when node dies and calling script

bschatz-swift committed:
Add a public network interface

bschatz-swift committed:
Add offline VIPs when dying and other cleanup

bschatz-swift committed:
Add UML to show transaction for failover

bschatz-swift committed:
New etcd import lines

bschatz-swift committed:
Update error handling

bschatz-swift committed:
Clear state of VGs online on local node if local node enters STARTING

bschatz-swift committed:
Have two VGs that are failed over. Use IP commands

Define two different VGs that must be failed over.

Use IP commands instead of ifconfig

Checkin two /etc/samba/smb-myTest* configuration files used for
prototype

bschatz-swift committed:
Code cleanup and node/vg offline

Cleanup the code.

Implement node offline and VG offline

Cleanup script which takes VIPs/smbd up and down

bschatz-swift committed:
Cleanup node code and names of node states

bschatz-swift committed:
Cleanup messages printed

bschatz-swift committed:
Add utility hacli to monitor/manage ha activities

bschatz-swift committed:
Add CLI to list nodes and vgs

bschatz-swift committed:
Remove unused code

bschatz-swift committed:
Update usage messages

bschatz-swift committed:
Offline changes

bschatz-swift committed:
Cleanup

bschatz-swift committed:
Reimplement offline - use rev to retrieve state

Handle offline coming from node code or via CLI

bschatz-swift committed:
Cleanup

bschatz-swift committed:
Node stop CLI

bschatz-swift committed:
Add online to CLI

bschatz-swift committed:
Fix bug shutting down node with no VGs online

bschatz-swift committed:
Latest files

Commit:
Update README file.

Add info about creating vboxnet5.

bschatz-swift committed:
Update smb .conf files for interface and fix race condition in shutdown

bschatz-swift committed:
Update state diagrams

Commit:
Bug fixes and cleanups and improved docs.

Minor bug fixes and improved(?) error messages.

Improved some documentation and added documention on pssh (parallel-shell).

bschatz-swift committed:
Add arping and set VG FAILED if offline failed.

bschatz-swift committed:
Simple script to loop failing over VGs

Commit:
Refactor API a bit, improve docs and error messages.

Add some new struct's to the API: vgInfo, NodeInfo, and ClusterInfo
(all include information about all volume groups, nodes, and single
cluster, respectively).

Replace several API functions that return 8 to 12 arguments with
the same function returning one of the above structs.

Introduce RevisionNumber data type and use it to old Raft version
numbers (with lots of casts between it and int64 at appropriate
places).

Improve some error messages in hacli.

Add a crucial sentence to the SETUP documentation describing the need
to create `/var/cache/samba`, without which smbd startup fails silently
(fortunately nmbd logs something).

bschatz-swift committed:
Correct path to hacli

bschatz-swift committed:
Pass netmask on the add of the VIP

bschatz-swift committed:
Remove debug messages and use VgInfo and VgNode in list functions

Commit:
Rename consensus.Struct --> consensus.EtcdConn

Pick a better name holding the information about our
connection to etcd and the status of outstanding operations.

Commit:
Minor improvements to SETUP documentation.

Add more info about configurating ssh and parallel-ssh.

Commit:
Refactor etcd interactions.

Change queries of the etcd database to return a slice of _comparisons_
and changes updates to items in the etcd datbase to return _operations_.
Actual updates to the etcd database consist of _conditionals_ and the
operations, where the operations are performed if, and only, all of the
conditionals (comparisons) are satisfied.  etcd guarantees that all of
the operations in an update are performed or none of them.

If proxyfs code concatenates the comparisons returned from all get
operations that make up a decision to change the database then it insures
the change is not made unless all of the prerequisites are still true.

Change the ProxyFS scheme for database storage to store all the
information about a Volume Group under a single key with the value
JSON encoded.  Refactor the interfaces to match. (The same change has
not yet been made for per node information, but it will be.)

One of the next steps is to make volumes (file systems) "first class"
citizens since, fundamentally, that is where the rubber meets the road.
Volumes are what get mounted, unmounted, failed, and are what the user
actually observes.  Changes to volume groups and, to a lesser extent,
nodes, only manifest themselves as changes to file systems.

Change the VgState type from an int to a string and discard code code
used to convert between the two.  make VgState a "first class" data type
in the shared database.

Change:
    VgInfo --> AllVgInfo
    NodeInfo --> AllNodeInfo

so we can use VgInfo and NodeInfo for information about an individual
node.  (AllVgInfo no longer exists as a type; its now represented as
a map[string]VgInfo.  but i think we really need to change this to
something that is based on UUID rather than user-assigned names.)

Update the SETUP documentation with a little more information about
parallel-ssh and the start of a "cheat sheet" on how to use etcdctl to
update the database state.

bschatz-swift committed:
Multiple node unit tests

Use the etcd framework for creating a multi node cluster.

At present, these tests do not attempt to verify any of the multiple
node state.  However, this commit adds the ability too.

bschatz-swift committed:
Used etcd integration test framework as basis for our unit tests

Use etcd integration framework as basis for our unit tests.  We no
longer need a running 3 node etcd cluster for the unit tests to pass.

Add flag to Struct signifying that we are running a unit test and
therefore set different arguments to vg_up_down script.

Change vg_up_down script to take different arguments mockUp and
mockDown.  In the future we will use these to simulate in a unit
test a failed up or failed down.

Change the API to be more Go like.`

Add a couple of functions in testutils for writing unit tests.  More
to come.

Commit:
Squash with previous consensus/vg.go checkin.

Delete unused code.

Commit:
Print tox version number.

Get tox to print its version number as part of `python-test` target.

bschatz-swift committed:
CWD to /tmp for unit tests

* etcd is unable to bind a Unix domain socket if the unit test is
running in a container.  Appears to be related to the file system used
to share the source.   The fix is to chdir to /tmp before creating the
test.
* remember starting current working directory so that we can execute the
proper up down script

bschatz-swift committed:
Integrate with etcd integration layer

Change unit tests to use etcd integration layer to simulate multiple
nodes

Add INITIAL as valid state when removing a VG

Commit:
Moved etcd interface code to etcdutil.go.

Refactor the code a bit more and move code that unpacks KeyValue
information returned by Get, Range, and etcd events into functions
in the new consensus/etcdutil.go file.

Probably the contents of watcher.go belong ther as well.

Commit:
Bug fixes.

Fix unpackKeyValue().  It turns that a KeyValue in an event can
have an empty value if the the event was for a key that was deleted.
Change unpackKeyValues() to notice when CreateVersion == 0 and return
a comparison function for the key that checks CreateVersion == 0 if
somebody wants to verify that key hasn't changed since it was deleted
(though this will not match correctly if the key is created and deleted
again before the check is made).

Fix a bug in unpackVgInfo() if is passed a VG that has been deleted
(probably only possible if the information is coming from an event).
In that case, don't try to unpack the (empty) value and just leave
the rest of the VgInfo with zero values.

Make up/down arguemnt to callUpDownScript() a real type, and map to
strings down where the script is called.

bschatz-swift committed:
Serialize shutdown and shutdown of HB goroutine

Serialize how HB is shutdown

bschatz-swift committed:
Signal that watcher is up AFTER calling Watch() and code cleanup

Wait to signal the WaitGroup until after we call Watch().

Clean up the code based on VS Code suggestions.

bschatz-swift committed:
Set AutoSyncInterval on client connections to 60 seconds

Setup list of endpoints used to be resynced once every 60 seconds.

Also add func Endpoints() to return list of end points

Commit:
Rename consensus/etcdutil.go --> etcd_util.go

Commit:
Write a pretty-printer for etcd transactions (conditions, succees actions
and "fail" actions).

Rewrite consensus/node.go to use a single key for each volume (the volume
state) and push the

Review the rest of the changes and whittle down the description later ...

Interim chekcin.

Write "generic" keyValue unpacking code in etcd_util.go.

The "node" and "vg" code still uses its own unpacking routines, but
i'm still removing bugs due to multiple keys.

Interim checkin.

Unit tests pass most of the time, but we need shutdown code for the
watchers (or at least wait until the nodes run through their state
changes).

Commit:
Interim checkin.

Trying to get the unit tests to run reliably.

Changes here cause the code to wait until the node is marked up
before proceeding which then caues all of the startVG code to
run which is turning up other problems ...

Commit:
Unit tests pass most of the time (all of the time?)

Shutdown the heartbeat goroutine before trying to exit.

Commit:
squash with previous -- minor "git vet" fix.

Commit:
Fix bug in dlm tests.

Commit:
Refactor the consensus code a bit.

Move the code that queries, updates, and deletes VG info and node info
into two new files:

	node.go --> node_etcd.go
	vg.go --> vg_etcd.go

and some generic helper routines in etcd_util.go

node_etcd.go and vg_etcd.go are relatively short at 180 lines each,
including comments, and very similair except for replacing "node" with
"vg" in a lot of data type and function names (try diff'ing them).
go doesn't have templates (a good thing), so i don't think its easily
possible to further squeeze out the common code.

Write a generic watcher goroutine wrapper called StartWatcher() and it
put it in etcd_util.go as well.  It provides channels to stop watchers,
collect error results, and wait for watchers to complete.

Rework watcher.go to use the new StartWatcher() routine and call into
node.go and vg.go to start the watchers that they need.

(It would be nice to have a generic routine to listen for responses to keys
that are watched but i haven't quite figured out what it should look like.)

Commit:
Add to the unit tests.

Remove some debug printfs.

bschatz-swift committed:
Unit test and example using the etcd wrappers

bschatz-swift committed:
Add a version which does similar steps without wrapper

bschatz-swift committed:
Add update to unit test

bschatz-swift committed:
Rename function to illustrated how used

bschatz-swift committed:
Remove unused file

bschatz-swift commented:
This is very old....
